### ISSUE TITLE
No uniform and buffer data shown in traces

### ISSUE THREAD

**Issue by OndrejSpanel** [apitrace/apitrace#365]
When inspecting traces with QApiTrace, all uniforms are shown as zero and no buffer data are shown. This happens with my traces, as well as a trace http://people.freedesktop.org/~jrfonseca/traces/furmark-1.8.2-svga.trace.

![image](https://cloud.githubusercontent.com/assets/6927223/8884921/1fc25324-325b-11e5-8f8d-85be6a5cf2cb.png)
![image](https://cloud.githubusercontent.com/assets/6927223/8884931/32fce76a-325b-11e5-8312-a51595b98057.png)

The screenshots above are captures at Frame 1, call 5928, which is glDrawElements. The framebuffer seems to be produced OK, the replay seems to be working, but uniform and buffer values seem missing or corrupt.

The build I am seeing this with was "apitrace-msvc-latest.7z 22-Jul-2015 11:31",

**Comment by jrfonseca** [comment]
It works fine here.  I tried on Windows with NVIDIA, and on Linux with
Intel.

I suspect that the problem may be specific to your OpenGL drivers.

Jose

**Comment by OndrejSpanel** [comment]
My Graphics card is ATI Radeon HD 5700, driver Catalyst 15.7 (that means July 2015).

OpenGL is reported as  6.14.10.13397.

Are the any log files, error output or debug output I could check so see why is ApiTrace not seeing the values and if perhaps this could be fixed somehow?

From all the OpenGL debugging tools this one looks the most promising to me so far, other I have tested did not work at all (crashed, did not capture anything) and I would really like to have it fully working.

I will definitely try it on another computer once I have the possibility.

**Comment by jrfonseca** [comment]
Try doing on the command line

     glretrace.exe -D 5928 furmark-1.8.2-svga.trace > nul:

and see if there's any messages shown on stderr.

**Comment by OndrejSpanel** [comment]
I see multiple repeated errors for my trace, but they do not seem to be related to uniforms or buffer, they are all like this:

> warning: message: high error 2311: glGetBooleanv parameter <active texture> has
an invalid value '8': must be less than 8 (GL_INVALID_VALUE)

With invalid value changing 8 to 15.

I see the same warnings for the "fur" trace I have linked to, and in addition also one error:


    3207: message: major api error 1001: glDisable parameter <cap> has an invalid enum '0x8910' (GL_INVALID_ENUM)
    3207: warning: glGetError(glDisable) = GL_INVALID_ENUM
    3917: warning: unsupported wglUseFontBitmapsA call
    3918: warning: unsupported wglUseFontBitmapsA call

**Comment by OndrejSpanel** [comment]
With a different GPU (nVidia GT 545) I can see uniform values OK. What I still do not see is the buffer values - perhaps I understood some functionality wrong? What I mean by "no buffer data" is the `Buffers` tab in the `Current State`. What kind of buffers is expected to be shown there?

On the nVidia card I can see the "Vertex Data" information on glBufferData calls, but not on `glDrawRangeElements` calls, where I expected them based on the screenshot showing a similar functionality for `glDrawElements` at http://zrusin.blogspot.cz/2011/04/apitrace.html.

**Comment by jrfonseca** [comment]
It does look like AMD drivers are busted somehow.  There's probably nothing that can be done.

Just one last request before I'm sure.  Could you run:

    apitrace.exe trace -o glretrace.trace glretrace.exe -D 5928 furmark-1.8.2-svga.trace > nul:

and provide the glretrace.trace file? That will enable me to verify with more confidence whether the drivers are busted or there's some bug in Apitrace.

> What kind of buffers is expected to be shown there?

Currently only transform feedback buffers appear there.  The hope is to eventually dump vertices/indices there too, but that hasn't been implemented yet.

> I expected them based on the screenshot showing a similar functionality for glDrawElements

That data appears can be seen only on the calls with blobs.  That particular call shows data on glDrawElements because a VBO was not used for the elements, hence the elements are passed in a blob to glDrawelements.

I admit it's not ideal.  Hence the hope to dump in on the buffers tab, but I can't give you any ETA.

**Comment by OndrejSpanel** [comment]
> provide the glretrace.trace file? 

Sure, [here it is](https://tresor.it/s#iBHVesqvtzjlCFhJ_x4yiw).

**Comment by jrfonseca** [comment]
Thanks.  I did look into it last week, and confirmed that gletrace seems to be doing everything as it was supposed to.

So I have little doubt that this is some sort of bug in AMD OpenGL drivers.  Most `glGet*` functions are not used by common OpenGL applications, so bugs in those functions can easily go undetected for quite some time, and I've seen this sort of bugs in the past.

That said, ApiTrace has to rely on OpenGL driver for this things.  Trying to cache state internally would be a hopeless battle: it would be an Herculian effort, and it would be susceptible to lots of bugs.  That is, it would introduce more problems than it would solve.

**Comment by jrfonseca** [comment]
Forgot to mention: if this is a system you often use to develop, you could consider to file an issue with AMD, as explained in the bottom of http://support.amd.com/en-us/kb-articles/Pages/AMDCatalyst15-7WINReleaseNotes.aspx  and provide `glretrace.exe -D 5928 furmark-1.8.2-svga.trace > nul:` as repro case.

**Comment by emeakyl** [comment]
I know this is an old issue, but I was able to reproduce it and I know where this happens.
Hash of the version I investigated: d83ab10a737d861f6c25a3727f6e28dbf876e62b ( Thu Jan 12 23:41:37 2017  )
My graphic card: AMD Radeon HD 7660D
Driver Packaging Version: 15.201.1151.1008-151104a-296217E
OpenGL Version: 6.14.10.13399

Everything goes well until glretrace dumps the code of the shaders. My ATI card supports ARB_get_program_binary, and so API trace makes a call in glstate_shaders.cpp to:
 glGetProgramiv(program, GL_PROGRAM_BINARY_LENGTH, &binaryLength);

and it resets the uniform values components to 0... Of course this is against the spec.
To make it work I add to comment/disable that block of code (nothing is lost since the binary format is not the one expected by the code)

I will try on another ATI card and fill a report to ATI.
